core/src/main/java/com/facebook/ktfmt/format/KotlinInputAstVisitor.kt

```
@@ -2395,9 +2400,12 @@ class KotlinInputAstVisitor(
     if (isEnabled) {
       open(plusIndent)
     }
-    block()
-    if (isEnabled) {
-      close()
+    try {
+      block()
+    } finally {
```

This seems like a good change, but I'm not sure how it's useful. Are there cases in the formatter where we swallow exceptions? I'd imagine that if `block()` throws, the whole formatter goes out the window.

---

It's not about exceptions, but control flow correctness. Because this is method is `inline`, calling `block()` can immediately return, skipping any subsequent statements. The fact that this is possible is a terrible language mistake.

---

Yikes, I had no idea. Good catch.

I just looked it up though and I think adding crossinline to the lambda would help, because it would prevent non-local returns.

I'd be okay with either solution, this one is certainly more flexible. I'd recommend adding a comment about it though.

---

The function could also not be `inline`. If nobody wants that control flow behaviour, there's no value in `inline` here.

---

> there's no value in `inline` here.

There's potentially a perf benefit. I just tried removing the inline, and it increased format time by 1.4%. Not a huge amount (we're talking 3.40ms -> 3.45ms per file, or an increase of 50 microseconds per file), but hey, it's not nothing :)

> I do want that behaviour in the PR though...

I mentioned in another comment, but I think it's possible to not need the early return. Am I understanding right that the early return is to avoid entering the `block` and therefore the linebreaks/indentation that come with it? If so you can just use the `isEnabled` parameter. Let me know if I'm missing something.